platform/source/mbed_os_timer.cpp

```diff
@@ -52,7 +52,13 @@ OsTimer *init_os_timer()
 #elif DEVICE_USTICKER
         os_timer = new (os_timer_data) OsTimer(get_us_ticker_data());
 #else
-        MBED_ERROR("OS timer not available");
+#if MBED_TRAP_ERRORS_ENABLED
```

Comment:
Thanks for updating the `MBED_ERROR` macro but it is safer to report the error even if MBED_TRAP_ERRORS_ENABLED is not set. This path is not valid. If there is an RTOS, then the OS timer needs to work. In bare metal, it's all fine until a timing function is used. A run-time error in this function is more useful than a null pointer exception later on.

Reply:
I have removed the `MBED_TRAP_ERRORS_ENABLED` if preprocessor so this error occurs if timing functions are used in baremetal regardless of error-masking configuration.

Reply:
> If there is an RTOS, then the OS timer needs to work.

In a case (e.g. RTOS) where a functionality is a strict requirement, a compile time error (e.g. `#error`) makes sense. Though we don't support RTOS (i.e. the full profile) on targets without timers to begin with I guess.

Reply:
There are a 4 cases here:

* RTOS in non-tickless mode. Uses SysTick peripheral, does not use `OsTimer`
* RTOS in tickless mode. Uses `OsTimer` to generate timebase from usticker or lpticker.
* Non-RTOS with timing calls. Uses `OsTimer`, inited on first use
* Non-RTOS with no timing (eg bootloader). `OsTimer` not needed; it and the ticker system should be linker excluded.

I agree a compile-time error makes sense, if there isn't one already, but this piece of code shouldn't really know the intricacies of who needs the OS timer. Not the place to put a bunch of "if RTOS and not tickless" defines. It's the lower layer.

You could arrange for `OsTimer` and `init_os_timer` to only be declared and defined if the timers were available, leading to compile errors on attempt to use, but those would not be terribly clear errors, and you'd have to add a whole bunch of ifdefs to existing code, meaning the problem cascades up somewhat.

With the updated APIs that are split between timing and non-timing (eg `Semaphore::acquire`/`try_acquire` versus `Semaphore::try_acquire_for`) you could have the same conditionality - `try_acquire_for` not available if no RTOS and no OsTimer.

But there are other calls that haven't had that split like `EventQueue::dispatch()` where `dispatch(-1)` or `dispatch(0)` is legal without the timer, but `dispatch(100)` isn't.   (Actually, due to implementation details, it currently isn't, but it could be). So they'd end up just being run-time checks. Those APIs should also be split - the split is what allows clean linker exclusion of timing code in bootloaders and other cut-down stuff.

Anyway, a happy medium would be that `init_os_timer` continues to exist and give a run-time error, to catch non-RTOS users going down a "needs timing" path. But there could be a define has matching the runtime condition like `OSTIMER_AVAILABLE` and the RTOS code could check that: `#if MBED_TICKLESS && !OSTIMER_AVAILABLE  / #error`.

I think that's a separate patch though - this PR is fine as-is.
That keeps the layering clean - the RTOS doesn't need to know how OSTimer is getting its time, and all the lpticker/usticker stuff, and OSTimer dosn't need to care about RTOSes.